Add emulated_roku docs (#6928)
home-assistant/home-assistant.github.io: source/_components/emulated_roku.markdown

---
layout: page
title: "Emulated Roku"
description: "Instructions on how to set up Emulated Roku within Home Assistant."
date: 2018-10-03 08:00
sidebar: true
comments: false
sharing: true
footer: true
logo: home-assistant.png
ha_category: Hub

> ha_release: 0.80.0

Reviewer: `0.80` has already been shipped. Please update this to the next major release.

ha_iot_class: "Local Push"
ha_qa_scale: internal
---

This component integrates an emulated Roku API into Home Assistant,  
so remotes such as Harmony and Android apps can connect to it through WiFi as it were a Roku player.  
Home Assistant will see key presses and app launches as Events, which you can use as triggers for automations.  
Multiple Roku servers may be started if you run out of buttons by specifying multiple server entries.  

> When using multiple servers, you can filter the roku USN which is specified via the `name` attribute.

Reviewer: ```suggestion
When using multiple servers, you can filter the Roku USN which is specified via the `name` attribute.
```

<p class='note'>  
Windows is not supported because Home Assistant uses `ProactorEventLoop` which does not support UDP sockets.
</p>  

## {% linkable_title Configuration %}

> configuration.yaml

Reviewer: Maybe remove this line, it is inconsistent with the rest of our documentation.

```yaml
# Example configuration.yaml entry
emulated_roku:
  servers:
    - name: hass roku
      listen_port: 8060
      host_ip: 192.168.1.150

>       advertise_ip: 10.0.0.10

Reviewer: We try to keep all configuration samples minimal. Thus, no optional requirement in the default sample. This helps a user to get started quickly by copy-&-paste the sample without worrying about optional parameters which they most likely not need. If required, insert a full configuration sample later that covers special setups or alike.

      advertise_port: 8070
      upnp_bind_multicast: True
```

{% configuration %}
name:

>   description: Name of the roku that will be displayed as USN in Harmony.

Reviewer: ```suggestion
  description: Name of the Roku that will be displayed as USN in Harmony.
```

  required: true
  type: string
listen_port:

>   description: The port the roku API will run on. This can be any free port on your system.

Reviewer: ```suggestion
  description: The port the Roku API will run on. This can be any free port on your system.
```

  required: true
  type: integer
host_ip:
  description: The IP address that your Home Assistant installation is running on. If you do not specify this option, the component will attempt to determine the IP address on its own.
  required: false
  type: string
advertise_ip:
  description: If you need to override the IP address used for UPnP discovery. (For example, using network isolation in Docker)
  required: false
  type: string
advertise_port:

>   description: If you need to specifically override the advertised UPnP port.

Reviewer: ```suggestion
  description: If you need to override the advertised UPnP port specifically.
```

Author: emulated_hue has the same description. I'd rather just remove the word, it's superfluous and the other property descriptions do not use it.

  required: false
  type: integer
upnp_bind_multicast:

>   description: Whether or not to bind the UPnP (SSDP) listener to the multicast address (239.255.255.250) or instead to the (unicast) host_ip address specified above (or automatically determined). The default is true, which will work for most situations.  In special circumstances, like running in a FreeBSD or FreeNAS jail, you may need to disable this.

Reviewer: ```suggestion
  description: Whether or not to bind the UPnP (SSDP) listener to the multicast address (239.255.255.250) or instead to the (unicast) host_ip address specified above (or automatically determined). The default is true, which will work in most situations.  In special circumstances, like running in a FreeBSD or FreeNAS jail, you may need to disable this.
```

  required: false
  type: boolean
  default: true
{% endconfiguration %}

> After starting up, you can check if the Emulated Roku is reachable at the specified ports on your hass instance (eg.: `http://192.168.1.101:8060/`)  

Reviewer: ```suggestion
After starting up, you can check if the Emulated Roku is reachable at the specified ports on your Home Assistant instance (eg.: `http://192.168.1.101:8060/`).
```

> If you change your advertised IP or ports you will have to re-add your Roku in Harmony.

Reviewer: ```suggestion
If you change your advertised IP or ports, you will have to re-add your Roku in Harmony.
```

## {% linkable_title Events %}

### {% linkable_title Event `roku_command` %}

> All roku commands are sent as `roku_command` events.

Reviewer: ```suggestion
All Roku commands are sent as `roku_command` events.
```

Field | Description
----- | -----------

> `source_name` | Name of the that sent the event roku. Not required for when using one Emulated Roku instance.

Reviewer: ```suggestion
`source_name` | Name of the that sent the event Roku. Not required for when using one Emulated Roku instance.
```

`type` | The type of the event that was called on the Roku API.
`key` | the code of the pressed key when the command `type` is `keypress`, `keyup` or `keydown`.
`app_id` | the id of the app that was launched when command `type` is `launch`.  

The available keys are listed here:
[Roku key codes](https://sdkdocs.roku.com/display/sdkdoc/External+Control+API#ExternalControlAPI-KeypressKeyValues)

## {% linkable_title Automations %}

> automations.yaml

Reviewer: Remove the file name and maybe write something like this:

```suggestion
The following is an example implementation of an automation:
```

```yaml
# Example automation
- id: amp_volume_up
  alias: Increase amplifier volume
  trigger:
  - platform: event
    event_type: roku_command
    event_data:
      source_name: hass roku
      type: keypress
      key: Fwd
  action:
  - service: media_player.volume_up
    entity_id: media_player.amplifier
```

## {% linkable_title Troubleshooting %}

Known limitations:

> * Some Android remotes send keyup / keydown events instead of keypress.

Reviewer: ```suggestion
* Some Android remotes send key up/down events instead of key presses.
```

---
layout: page
title: "Emulated Roku"
description: "Instructions on how to set up Emulated Roku within Home Assistant."
date: 2019-01-10 08:00
sidebar: true
comments: false
sharing: true
footer: true
logo: home-assistant.png
ha_category: Hub
ha_release: 0.86.0
ha_iot_class: "Local Push"

> ha_qa_scale: internal

Reviewer 2: I don't think this should be classified as internal. Is there a precedence?

Author: Yeah, probably not. Removed.
